submariner/multiple-cable-drivers.md

```diff
@@ -0,0 +1,415 @@
+cluster set. For example, a cluster set could contain three clusters, two that are on-premise and the other
+in the cloud. Connections to the cloud cluster will need to use encryption while the on-premise clusters can use
+unencrypted connections, as they’re connected by a private link, to avoid the performance overhead of encryption.
+This could be achieved using either:
+
+- Different cable driver implementations or
+- The same cable driver implementation with differing configurations.
+
+For other use cases, it makes sense to offer finer-grained connectivity options. One such use case is a client-server
+scenario where multiple client clusters need to access a server cluster but the user does not want the client clusters
+to connect to one another. Finer-grained connectivity could also improve overall scalability where full mesh is not
+required.
+
+## Proposal
+
+This enhancement proposes a mechanism to allow users to specify their intent as to which clusters to connect and how to
+connect them. The default would still be a full mesh with a homogeneous cable driver.
+
+The standard mechanism in Kubernetes to specify intent is by creating a Kubernetes API resource whose effect is eventually
+realized into the cluster's desired state. To extend the Kubernetes API, one could either use the generic Kubernetes
+`ConfigMap` or use a custom resource definition (CRD). There's pros and cons to each but a CRD seems more suitable for
+this enhancement.
+
+This enhancement will propose a single API for finer-grained connectivity and multiple cable driver support.
+
+### Finer-grained connectivity and multiple cable driver API
+
+This following CRD is proposed:
+
+```Go
+type ClusterConnectionPolicies struct {
+    metav1.TypeMeta `json:",inline"`
+
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec ClusterConnectionPolicy `json:"spec"`
+}
+
+type ClusterConnectionPolicySpec struct {
+    // LeftClusterSelector identifies the Cluster resources representing the clusters on one end of a connection. An empty
+    // selector indicates wildcard, ie matches any cluster.
+    // Not used in combination with ClusterSelector.
+    LeftClusterSelector metav1.LabelSelector `json:"leftClusterSelector,omitempty"`
+
+    // RightClusterSelector identifies the Cluster resources representing the clusters on the other end of a connection.
+    // An empty selector indicates wildcard, ie matches any cluster.
+    // Not used in combination with ClusterSelector.
+    RightClusterSelector metav1.LabelSelector `json:"rightClusterSelector,omitempty"`
+
+    // ClusterSelector identifies the Cluster resources representing a subset of clusters.
+    // Not used in combination with LeftClusterSelector or RightClusterSelector.
+    // An empty selector indicates wildcard, ie matches any cluster.
+    ClusterSelector metav1.LabelSelector `json:"clusterSelector,omitempty"`
```

Comment:
Why is there this option?

Reply:
Just an option, if a bunch of clusters share a label, e.g. in the attached example it's usage is shown on L262... to connect all the clusters labelled cloud in a full mesh fashion...

Reply:
I was also a little confused by this option. Can't that be accomplished by:

```
leftClusterSelector:
    location: cloud
rightClusterSelector:
    location: cloud
```